Event 436 part 2

jayakrishnanv opened the pull request:
@hlbraddock @OzzieOrca @jcsaplar Implemented part 2 of EVENT-436. Please test and merge.

jayakrishnanv commented:
@hlbraddock Rebased from master

hlbraddock commented:
@jayakrishnanv Please rebase against master. This will be the next branch we merge.

hlbraddock commented:
@OzzieOrca please review.

hlbraddock commented:
@jayakrishnanv You might want to wait until after this is reviewed before rebasing.

hlbraddock commented:
@jayakrishnanv This is a really nice feature. +1

1. Can you reword : `This Rules section allows for conditional enforcement of any answer marked 'Force Selection'.`
2. Can you hide this Rules section altogether if no answer has 'Force Selection'?

jayakrishnanv commented:
@hlbraddock We have changed the information text in force selection rule block and made the rule selection hidden altogether if no answer has 'Force Selection'

hlbraddock commented:
@OzzieOrca Mr. Orca. Are you happy with this PR?

OzzieOrca commented on app/views/components/blockEditor.html:
```diff
@@ -204,19 +204,86 @@
           </div>
         </div>
+
+        <div ng-if="block.type === 'checkboxQuestion' && disableForceSelectionRule() === false">
```
I couldn't figure out how to get this to show up. Maybe I just don't know how to use it... I have a checkbox question with some force selections.

jayakrishnanv replied:
@OzzieOrca The new rules section is hidden by default. It becomes visible beneath the existing rules section only when the administrator does the following:

1. Check an option, which brings up a new check box "Force Selection" to the right.
2. Check the "Force Selection" check box.

  You may have missed it, since it is visible only if the edit panel is open.

OzzieOrca replied:
I feel dumb. I think I got redirected to stage and didn't realize it. Sorry. I see it now.

hlbraddock replied:
@jayakrishnanv I think that text is fine.

OzzieOrca commented on app/views/components/blockEditor.html:
```diff
@@ -204,19 +204,86 @@
+                <button class="btn btn-success" ng-click="addForceRule()"><i class="fa fa-plus"></i> <translate>Add Rule</translate></button>
+              </div>
+              <div class="col-md-8 rule-message-alignment">
+                <span class="pull-left" translate>Show question if:</span>
```
This needs to be something related to force selections. `Force selections if` or something similar.

OzzieOrca replied:
I think these 2 rule sections could use some headings. Maybe `Question visibility rules` and `Force selection rules`. Idk exactly what they should be. @hlbraddock do you have any thoughts?

jayakrishnanv replied:
@OzzieOrca I agree with changing to `Force selections if`.

jayakrishnanv replied:
@OzzieOrca Aren't the existing information headers enough?

`Rules are a great way to hide questions based on the registrant's answers to other questions. All rules created must be met before the question is displayed. Rules can depend on questions of these types that are above this question.` for normal visibility rules

and

`This Rules section allows for conditional enforcement of any answer marked 'Force Selection'. Rules can depend on questions of these types that are above this question.` for force selection rules?

OzzieOrca replied:
I think it is just hard to quickly scan and know what section is which. Currently it looks like:
![screen shot 2016-10-10 at 11 18 05 pm](https://cloud.githubusercontent.com/assets/756501/19260607/d3b91e0c-8f3f-11e6-9a2f-e868f430b67c.png)

I was thinking having some bolder title would be nice. Smething like this maybe: (Using `h4`. Maybe the spacing could be smaller.)
![screen shot 2016-10-10 at 11 05 17 pm](https://cloud.githubusercontent.com/assets/756501/19260490/d88fbb30-8f3e-11e6-97a5-667b323d4092.png)

I was also wondering if bringing the radios up to the top would work since they are really what the action here is. Idk that I've got it looking exactly how I want it but here's an idea:
![screen shot 2016-10-10 at 11 46 34 pm](https://cloud.githubusercontent.com/assets/756501/19261102/d0ea2ece-8f43-11e6-953c-474de88038a9.png)

I was trying to consolidate the help message. Maybe something like this would be good:

> Rules are a great way to show questions or force selections based on the registrant's answers to other questions. Rules can depend on questions of these types that are asked before this question.

You could hide the `or force selections` part if it is not a checkbox question.

I was just playing around with it in the chrome inspector but here's my code from moving it around.
```html
<div class="tab-pane ng-scope active" ng-repeat="tab in tabs" ng-class="{active: tab.active}" tab-content-transclude="tab">
    
        <p class="text-muted ng-scope"><i class="fa fa-info-circle"></i>Rules are a great way to show questions or force selections based on the registrant's answers to other questions. Rules can depend on questions of <a href="" tooltip="Multiple Choice, Dropdown, Number, Date, Gender or Year in School" translate="">these types</a> that are asked before this question.</p>
        <div class="clearfix rule-message-alignment">
              <h4 class="pull-left" translate=""><span class="ng-scope">Show question if:</span></h4>
              <!-- ngRepeat: choice in ['AND','OR'] --><div ng-repeat="choice in ['AND','OR']" class="pull-left margin-left-10 radio padding-top-0 ng-scope">
                <label>
                  <input type="radio" show-errors="" ng-model="block.content.ruleoperand" ng-value="choice" class="ng-pristine ng-untouched ng-valid" name="224" value="AND">
                  <!-- ngIf: choice==='AND' --><strong ng-if="choice==='AND'" translate="" class="ng-scope"><span class="ng-scope">All rules are met</span></strong><!-- end ngIf: choice==='AND' -->
                  <!-- ngIf: choice==='OR' -->
                </label>
              </div><!-- end ngRepeat: choice in ['AND','OR'] --><div ng-repeat="choice in ['AND','OR']" class="pull-left margin-left-10 radio padding-top-0 ng-scope">
                <label>
                  <input type="radio" show-errors="" ng-model="block.content.ruleoperand" ng-value="choice" class="ng-pristine ng-untouched ng-valid" name="226" value="OR">
                  <!-- ngIf: choice==='AND' -->
                  <!-- ngIf: choice==='OR' --><strong ng-if="choice==='OR'" translate="" class="ng-scope"><span class="ng-scope">Any rules are met</span></strong><!-- end ngIf: choice==='OR' -->
                </label>
              </div><!-- end ngRepeat: choice in ['AND','OR'] -->
            <button class="btn btn-success pull-right" ng-click="addRule()"><i class="fa fa-plus"></i> <translate><span class="ng-scope">Add Rule</span></translate></button></div>
```

OzzieOrca replied:
No, I was having more of a discussion of ideas. What do you and @hlbraddock think? I think this change will only need to be made in blockEditor.html and should affect all questions from there. I guess this is less important for the other questions but I still think it would be a good change since it highlights the action instead of burying it beneath a large block of text the same size as it and buttons.

jayakrishnanv replied:
@OzzieOrca Yes, it's a common component, so you change it once, it reflects across all questions. I'll leave it to you and @hlbraddock to finalise the design.

hlbraddock replied:
@jayakrishnanv @OzzieOrca 's I like the proposed ideas.  Let's implement these:

. Display just one single consolidated help message (as shown in example above)
. Add Title for each Rule section "Show Question Rules" and "Force Selection Rules"
. Use radio text "Show question if" and "Force selections if"
. Right justify 'Add Rule' buttons

jayakrishnanv replied:
@hlbraddock @OzzieOrca @jcsaplar We have made the changes as per your suggestion, please test:

front-end - https://github.com/CruGlobal/conf-registration-web/pull/510
back-end - https://github.com/CruGlobal/ert-api/pull/679

jayakrishnanv replied:
@hlbraddock @OzzieOrca @jcsaplar Youmay have to revert the merge done on stage for #506 before merging #510

OzzieOrca commented:
I've also been getting this error after creating a checkbox and adding a single possible answer. I haven't touched default answers, force selection, or rules.
Deleting the question makes my event save correctly again but adding a checkbox and a possible answer again results in that error again.
```
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "additionalRules" (class org.cru.crs.api.model.Block), not marked as ignorable (13 known properties: "profileType", "position", "pageId", "registrantTypes", "exportFieldTitle", "expenseType", "rules", "title", "content", "type", "id", "adminOnly", "required"]) at [Source: java.io.ByteArrayInputStream@23bda9de; line: 1, column: 5689] (through reference chain: org.cru.crs.api.model.Conference["registrationPages"]->java.util.ArrayList[0]->org.cru.crs.api.model.Page["blocks"]->java.util.ArrayList[14]->org.cru.crs.api.model.Block["additionalRules"])
```

jayakrishnanv commented:
@OzzieOrca I'm not able to recreate the issue you mentioned. Can you explain the steps you followed? I mean is it due to any other questions you added before this check box question?

OzzieOrca commented:
Steps to reproduce:
1. Create a new event
2. Drag a checkbox question to be the last question
3. Add "test" as a possible answer

I immediately get "Error Update failed" and a network request failure with that error message.

Here's my curl request if that is of any interest to you:
```bash
curl 'https://api.stage.eventregistrationtool.com/eventhub-api/rest/conferences/90ecb84a-a61d-4d79-8605-4169b627e9f2' -X PUT -H 'Authorization: 2a7beb8b83d6d9ae8b650b5d102c631ecf05cf0b' -H 'Origin: http://localhost:9000' -H 'Accept-Encoding: gzip, deflate, sdch, br' -H 'Accept-Language: en-US,en;q=0.8' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.143 Safari/537.36' -H 'Content-Type: application/json;charset=UTF-8' -H 'Accept: application/json, text/plain, */*' -H 'Referer: http://localhost:9000/' -H 'Cookie: JSESSIONID=JNvwfQ30baZIoU3onyBRrLmnA1DGygcXTPogr20f.235520cd536f; _ga=GA1.2.1080765212.1473295479' -H 'Connection: keep-alive' --data-binary '{"id":"90ecb84a-a61d-4d79-8605-4169b627e9f2","name":"Broken checkbox question","description":null,"abbreviation":null,"registrationPages":[{"id":"9104779a-d2c4-4c2c-b98c-a61944685e2f","conferenceId":"90ecb84a-a61d-4d79-8605-4169b627e9f2","title":"Your Information","position":0,"blocks":[{"id":"d065909c-2261-4298-b350-caaacc916b87","pageId":"9104779a-d2c4-4c2c-b98c-a61944685e2f","title":"Name","exportFieldTitle":null,"type":"nameQuestion","required":true,"position":0,"adminOnly":false,"content":{"default":"","ruleoperand":"AND","forceSelections":{},"forceSelectionRuleOperand":"AND"},"profileType":"NAME","registrantTypes":[],"rules":[],"expenseType":null},{"id":"fce2a8d8-ebfe-426f-afa9-c43dbcfdac63","pageId":"9104779a-d2c4-4c2c-b98c-a61944685e2f","title":"Email","exportFieldTitle":null,"type":"emailQuestion","required":true,"position":1,"adminOnly":false,"content":{"default":"","ruleoperand":"AND","forceSelections":{},"forceSelectionRuleOperand":"AND"},"profileType":"EMAIL","registrantTypes":[],"rules":[],"expenseType":null},{"id":"17c69936-6428-47f5-ada4-ac2375510a11","content":{"default":"","ruleoperand":"AND","forceSelections":{},"forceSelectionRuleOperand":"AND","choices":[{"value":"test","desc":""}]},"pageId":"9104779a-d2c4-4c2c-b98c-a61944685e2f","required":false,"title":"Checkbox Question","type":"checkboxQuestion","profileType":null,"registrantTypes":[],"rules":[],"additionalRules":[]}]}],"registrantTypes":[{"id":"fac2c6ea-b0c2-4e84-aa5e-a23b1a59b162","conferenceId":"90ecb84a-a61d-4d79-8605-4169b627e9f2","name":"Default","hidden":false,"createdTimestamp":"2016-10-11T06:47:46.590Z","lastUpdatedTimestamp":"2016-10-11T06:47:46.590Z","cost":0,"calculatedCurrentCost":0,"minimumDeposit":null,"position":0,"customConfirmationEmailText":null,"description":null,"groupSubRegistrantType":false,"allowGroupRegistrations":false,"familyStatus":"DISABLED","numberSlotsLimit":0,"useLimit":false,"availableSlots":0,"acceptCreditCards":false,"acceptTransfers":false,"acceptScholarships":false,"acceptChecks":false,"acceptPayOnSite":false,"defaultTypeKey":null,"earlyRegistrationDiscounts":[]}],"promotions":[],"eventStartTime":"2016-10-24 23:47:46","eventEndTime":"2016-10-30 23:47:46","registrationStartTime":"2016-10-10 23:47:46","registrationEndTime":"2016-10-24 23:47:46","eventTimezone":"America/New_York","registrationOpen":true,"contactPersonName":"Scotty Waggoner","contactPersonEmail":"scotty.waggoner@cru.org","contactPersonPhone":"12093224081","contactWebsite":null,"registrationCompleteRedirect":null,"locationName":null,"locationAddress":null,"locationCity":null,"locationState":null,"locationZipCode":null,"requireLogin":false,"archived":false,"paymentGatewayType":"AUTHORIZE_NET","paymentGatewayId":null,"paymentGatewayKey":null,"paymentGatewayKeySaved":false,"registrationCount":0,"completedRegistrationCount":0,"customPaymentEmailText":null,"rideshareEnabled":false,"rideshareEmailContent":null,"allowEditRegistrationAfterComplete":true,"checkPayableTo":null,"checkMailingAddress":null,"checkMailingCity":null,"checkMailingState":null,"checkMailingZip":null,"businessUnit":null,"operatingUnit":null,"department":null,"projectId":null,"accountNumber":null,"glAccount":null,"loggedInUserPermissionLevel":null,"cruEvent":false}' --compressed
```

jayakrishnanv commented:
@OzzieOrca Have you deployed the back-end changes related to this task? We have added a new table and modified the back-end code for this task. Please deploy this branch https://github.com/CruGlobal/ert-api/pull/678

OzzieOrca commented:
Ah that would be the reason. I've never run the API locally. I assumed the changes would be on stage. Ignore that then as long as it works when the API and front-end match.

jcsaplar commented:
@OzzieOrca @jayakrishnanv @hlbraddock Both the front and backend are pushed to stage and working You should be able to use stage now. Let me know if you have any other changes.

OzzieOrca commented:
Did you make a new PR for this again? Is it now in https://github.com/CruGlobal/conf-registration-web/pull/510? I think it would be good to at least tag the old one if you have to create a new PR. Should this be closed now?

hlbraddock commented:
Correct.

This should be closed

> On Oct 14, 2016, at 7:57 PM, Scotty Waggoner <notifications@github.com> wrote:
> 
> Did you make a new PR for this again? Is it now in #510? I think it would be good to at least tag the old one if you have to create a new PR. Should this be closed now?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hlbraddock commented:
Replaced by https://github.com/CruGlobal/conf-registration-web/pull/510
